git-config/src/parse.rs

```diff
@@ -0,0 +1,139 @@
+use crate::{file, spanned};
+use dangerous::{Bytes, BytesReader, Error, Expected, Input};
+
+fn config(bytes: &[u8]) -> Result<Vec<file::Token>, Expected<'_>> {
+    fn config<'i, E>(input: Bytes<'i>, r: &mut BytesReader<'i, E>) -> Result<Vec<file::Token>, E>
+    where
+        E: Error<'i>,
+    {
+        let mut tokens = Vec::new();
+        skip_whitespace_or_comment(r, ConsumeTo::NextToken).map(|section| {
+            tokens.push(spanned::Comment(
+                dangerous::input(section)
+                    .span_of(&input)
+                    .expect("range contained")
+                    .into(),
+            ))
+        });
+        unimplemented!("sections and values");
+    }
+    let input = dangerous::input(bytes);
+    input.read_all(|r| config(dangerous::input(bytes), r))
+}
+
+enum ConsumeTo {
+    NextToken,
+    EndOfLine,
+}
+
+#[must_use]
+fn skip_whitespace_or_comment<'a, E>(r: &mut BytesReader<'a, E>, to_where: ConsumeTo) -> Option<&'a [u8]> {
+    fn skip_whitespace_or_comment<E>(r: &mut BytesReader<'_, E>, to_where: ConsumeTo) {
+        fn skip_comment<E>(r: &mut BytesReader<'_, E>) -> usize {
+            if r.peek_eq(b'#') {
+                r.take_while(|c| c != b'\n').len()
+            } else {
+                0
+            }
+        }
+
+        let (mut last, mut current) = (0, 0);
+        loop {
+            current += skip_comment(r);
+            current += r
+                .take_while(|c| {
+                    let iwb = c.is_ascii_whitespace();
+                    iwb && match to_where {
+                        ConsumeTo::NextToken => true,
+                        ConsumeTo::EndOfLine => c != b'\n',
+                    }
+                })
+                .len();
+            if last == current {
+                break;
+            }
+            last = current;
+        }
+    }
+    let parsed = r
+        .take_consumed(|r| skip_whitespace_or_comment(r, to_where))
+        .as_dangerous();
```

Comment:
Here I thought there could be an `as_dangerous_opt()`, as I specifically avoided doing `as_dangerous_non_empty().ok()` probably due to a very premature optimization. Generally I like the `_opt()` versions of methods a lot.